ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +175,420 @@ let order_f (_, vm_rec) =
+             which has running VMs from the same group is no less than 2, \
+             continue to plan for the next one." ;
+          (acc_mapping, acc_pool_state)
+    )
+    ([], pool_state) anti_affinity_vms
+  |> fst
+
+let anti_affinity_vms_increasing ~__context vms =
+  vms
+  |> List.filter_map (fun (vm, vm_size) ->
+         match Xapi_vm_helpers.vm_has_anti_affinity ~__context ~vm with
+         | Some (`AntiAffinity group) ->
+             Some (vm, vm_size, group)
+         | _ ->
+             None
+     )
+  |> List.sort (fun (_, a, _) (_, b, _) -> compare a b)
+
+(* Return an evacuation plan respecting VM anti-affinity rules: it is done in 3
+   phases:
+   1. Try to get a "spread evenly" plan for anti-affinity VMs, and then a
+      binpack plan for the rest of VMs. Done if every VM got planned, otherwise
+      continue.
+   2. Try to get a "no breach" plan for anti-affinity VMs, and then a binpack
+      plan for the rest of VMs. Done if every VM got planned, otherwise
+      continue.
+   3. Carry out a binpack plan ignoring VM anti-affinity. *)
+let compute_anti_affinity_evac_plan ~__context total_hosts hosts vms =
   let config =
     {Binpack.hosts; vms; placement= []; total_hosts; num_failures= 1}
   in
   Binpack.check_configuration config ;
-  debug "Planning configuration for offline agile VMs = %s"
-    (Binpack.string_of_configuration
-       (fun x ->
-         Printf.sprintf "%s (%s)" (Ref.short_string_of x)
-           (Db.Host.get_hostname ~__context ~self:x)
-       )
-       (fun x ->
-         Printf.sprintf "%s (%s)" (Ref.short_string_of x)
-           (Db.VM.get_name_label ~__context ~self:x)
-       )
-       config
-    ) ;
+
+  let binpack_plan ~__context config vms =
+    debug "Binpack planning configuration = %s"
+      (Binpack.string_of_configuration
+         (fun x ->
+           Printf.sprintf "%s (%s)" (Ref.short_string_of x)
+             (Db.Host.get_name_label ~__context ~self:x)
+         )
+         (fun x ->
+           Printf.sprintf "%s (%s)" (Ref.short_string_of x)
+             (Db.VM.get_name_label ~__context ~self:x)
+         )
+         config
+      ) ;
+    debug "VMs to attempt to evacuate: [ %s ]"
+      (String.concat "; "
+         (vms
+         |> List.map (fun (self, _) ->
+                Printf.sprintf "%s (%s)" (Ref.short_string_of self)
+                  (Db.VM.get_name_label ~__context ~self)
+            )
+         )
+      ) ;
+    let h = Binpack.choose_heuristic config in
+    h.Binpack.get_specific_plan config (List.map fst vms)
+  in
+
+  let binpack_after_plan_applied = function
+    | [] ->
+        None
+    | plan -> (
+        debug "Binpack for the rest VMs" ;
+        let config_after_plan_applied = Binpack.apply_plan config plan in
+        let vms_to_plan =
+          let planned_vm = List.map fst plan in
+          vms |> List.filter (fun (vm, _) -> not (List.mem vm planned_vm))
```

Comment:
It's not efficient to do this searching in a list. The VMs to be planned:
1. spread-evenly: always empty
2. no-breach: can be returned from `compute_no_breach_plan`.
Put them together, the returned plan from both cases above can be a tuple `(planned_aff_vms, unplanned_aff_vms)`
And the to-be-planned VMs will be `non_aff_vms @ unplanned_aff_vms`, where the `non_aff_vms` can be returned from `anti_affinity_vms_increasing`/`is_aff_vm` below in which there will be a partition on the `vms`.